Introduce Docker compose file & setup-wizard e2e test

juliaamosova opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

This PR includes the following:

1) New `docker-compose.yaml` and `Dockerfile` in order to run e2e tests locally using Docker;
2) New `@jest/test-sequencer` package in order to run tests in a particular order;
3) New `activate-and-setup/setup-wizard.test.js` - new e2e test that will run before all other tests;
4) New `test:e2e-dev` script for running e2e tests in `dev` mode (headless mode off and SloMo mode on);
5) New config file `test:e2e.json` to pull variables from, needed to run e2e tests.
6) Travis implementation for running e2e tests with Docker. 

More details on each point:

1) **Docker**: this PR adds Dockerfiles needed to build a test site locally using Docker. An [official WordPress Docker image](https://github.com/docker-library/docs/blob/master/wordpress/README.md#-via-docker-stack-deploy-or-docker-compose) is used to build the site.  Once the site using the WP Docker image is built, the current WooCommerce dev branch is being copied to the `plugins` folder of that newly built test site. No WooCommerce Docker image is being built or needed. 

2) **Jest test sequencer**: by default, jest runs tests ordered by the time it takes to run the test (the test that takes longer to run will be run first, the test that takes less time to run will run last). Jest sequencer introduces tools that can be used to specify the order in which the tests are being run. In our case, they are being run in alphabetical order of the directories where tests are located. This way, tests in the new directory `activate-and-setup` will run first. 

3) **Setup Wizard E2E test**: this test (located in `activate-and-setup` directory) will run before all other tests. This will allow making sure that WooCommerce is activated on the site and for the setup wizard to be completed on a brand new install of WooCommerce + it will also test Setup Wizard flow. 

4) **`test:e2e-dev` script**: this allows to run `npm run test:e2e-dev` in order to run e2e test suite in dev mode. Previously, we had to add `--dev` flag to the `test:e2e` script. 

5) **`test:e2e.json` config file**: needed so that we won't need to set test site URL, admin and customer details manually anymore. 

6) **Travis**: we can't test the Travis implementation in this PR since this branch was not made of the master. We need to merge this PR to our migration branch (https://github.com/woocommerce/woocommerce/pull/24218) which is of the master to see if the Travis implementation works and how tests run there. 

### How to test the changes in this Pull Request:

1. Install Docker Desktop if you don't have it installed:

- [Docker Desktop for Mac](https://docs.docker.com/docker-for-mac/install/)
- [Docker Desktop for Windows](https://docs.docker.com/docker-for-windows/install/)

Once installed, you should see `Docker Desktop is running` message with the green light next to it (Docker Desktop for Mac):

![docker](https://user-images.githubusercontent.com/19143190/69484051-891ac300-0e26-11ea-84ac-93a8d235173e.jpg)

**Note, that if you install docker through other methods such as homebrew, for example, your steps to set it up will be different. The commands listed in steps below may also vary.** 

2. Checkout the branch of this PR to test and stay on this branch. 

Note that by default, the local test site will be built to access on `http://localhost:8084/`. If you need to change the port (from `8084` to another one), open `docker-compose.yaml` file and change it on lines 23 and 44. 

3. Run `nmp install`

4. Run `composer install --no-dev`

5. Run `npm run build`

6. Run the following command to build the test site using Docker: `docker-compose up` and watch the site being built. Note that it may take a few minutes the first time you do that. 
7. Open new terminal window and `cd` to the current branch again.
8. Run the following command to make sure the containers were built and running: `docker ps`. You should see the 2 following containers: 

- `woocommerce_wordpress-woocomerce-dev`
- `mariadb:10.4`

9. Navigate to `http://localhost:8084/` to view the newly built site. 

10. Since we specify the site URL and create admin and customer users in docker-compose file, there is `./tests/e2e-tests/config/default.json` file that contains the variables needed to run the test:

```
{
  "url": "http://localhost:8084/",
  "users": {
    "admin": {
      "username": "admin",
      "password": "password"
    },
    "customer": {
      "username": "customer",
      "password": "password"
    }
  }
}
```

If you changed either site URL or users details in the docker-compose file, you'd need to copy the content of the `default.json`, paste it to `test:e2e.json` and edit it further there to match your own setup. 

11. We can now run the tests by running: 

- `npm run test:e2e` (headless)
- `npm run test:e2e-dev` (dev mode: non-headless with SloMo)

12. Make sure all tests are passing. 

**Note that after you run the test suite for the first time and complete the Setup Wizard**, you will need to skip the Setup Wizard next time you want to run test suite again on the same install. To skip the tests, use `.only` in the relevant test entry to specify the tests that you do want to run. 

For example, in order to skip Setup Wizard tests, add `.only` to the login and activation tests as follows in the `setup-wizard.test.js`:

```
it.only( 'Can login', async () => {}
```

```
it.only( 'Can make sure WooCommerce is activated. If not, activate it', async () => {}
```

As a result, when you run `setup-wizard.test.js`, only the login and activate tests will run. The rest will be skipped. You should see the following in the terminal:

```
 PASS  tests/e2e-tests/specs/activate-and-setup/setup-wizard.test.js (11.927s)
  Store owner can login and make sure WooCommerce is activated
    ✓ Can login (7189ms)
    ✓ Can make sure WooCommerce is activated. If not, activate it (1187ms)
  Store owner can go through store Setup Wizard
    ○ skipped Can start Setup Wizard
    ○ skipped Can fill out Store setup details
    ○ skipped Can fill out Payment details
    ○ skipped Can fill out Shipping details
    ○ skipped Can fill out Recommended details
    ○ skipped Can skip Activate Jetpack section
    ○ skipped Can finish Setup Wizard - Ready! section
  Store owner can finish initial store setup
    ○ skipped Can enable tax rates and calculations
    ○ skipped Can configure permalink settings
```

<hr>

**TO STOP & RESTART DOCKER:**

- Press `Ctrl+C` in the terminal window where the containers are running 
- Stop the container(s) using the following command: `docker-compose down`
- Restart the containers using the following command: `docker-compose up`

**TO STOP DOCKER & DO A CLEAN RESTART:**

Steps below will allow building a brand new site with a clean DB and no data as it was built initially:

- Press `Ctrl+C` in the terminal window where the containers are running 
- Stop the container(s) and delete all volumes using the following command: `docker-compose down -v`
- Restart the containers using the following command: `docker-compose up --build`

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

> Enter a summary of all changes on this Pull Request. This will appear in the changelog if accepted.

juliaamosova committed:
Introduce Docker compose file & setup-wizard e2e test

juliaamosova committed:
Resolving conflict with feature branch

juliaamosova committed:
Finish remaining work on the setup-wizard e2e test

juliaamosova committed:
Add config file to store variables needed to run e2e tests

juliaamosova committed:
Remove test:e2e.json file from the code base

juliaamosova committed:
Update admin and customer user details in docker compose file

juliaamosova committed:
Fix the way usage tracking pop-up window is handled

kloon commented on Dockerfile:
```diff
@@ -0,0 +1,3 @@
+FROM wordpress:5.3
```
I think we can use wordpress:latest here to not be bound by a specific version and to make it future proof without having to update the file every time a new version of WP is released.

juliaamosova replied:
I wasn't sure if `wordpress:latest` always pull the latest stable version or if it could be the nightly build which is why I decided to specify the version. I see, however, that at least for now, when using `wordpress:latest`, 5.3 is being installed on the site. I also see that we use `wordpress:latest` for running unit tests for example. I think it is OK to change it to `wordpress:latest` which is what I did.

kloon commented on docker-compose.yaml:
```diff
@@ -0,0 +1,53 @@
+version: '3.1'
```
Any reason for using version 3.1 here? Latest version of docker compose is 3.7 which supports Docker Engine 18.06.0 and up. I believe the latest version of Docker Engine is 19.03 atm.

juliaamosova replied:
No specific reason. I used the default `docker-compose.yaml` from the [official WordPress docker image documentation](https://github.com/docker-library/docs/blob/master/wordpress/README.md#-via-docker-stack-deploy-or-docker-compose) to start working on the setup and didn't think to change the version. I updated it to 3.7 now. Thanks for catching it!

kloon commented on docker-compose.yaml:
```diff
@@ -0,0 +1,53 @@
+services:
+
+  db:
+    image: mysql:5.7
```
Instead of mysql we could also use mariadb which is a dropin for mysql which most hosting companies use.

juliaamosova replied:
Sounds good to me. I updated the db image to use mariadb.

kloon commented on docker-compose.yaml:
```diff
@@ -0,0 +1,53 @@
+    command: >
+      /bin/sh -c '
+      wp core install --url=http://localhost:8084 --title="WooCommerce Core E2E Test Suite" --admin_user=admin --admin_password=password --admin_email=admin@woocommercecoree2etestsuite.com --path=/var/www/html --skip-email;
+      wp plugin activate woocommerce;
```
I am wondering since the source is coming from github it might be missing the build dependencies, so before activating we might need to run `composer install --no-dev` and `npm install`

juliaamosova replied:
That's right, it does miss that. Thanks for catching it! I am not sure it needs to be done within a Docker container. 

When running tests locally, we can run it before we run the tests. I updated the instructions on how to test this PR locally with these 2 commands. 

When it comes to Travis, I see that we run these commands in the `before_script` section of `travis.yml`:

`composer install --no-dev`:

https://github.com/woocommerce/woocommerce/blob/master/.travis.yml#L66

`npm install` is currently being run here within `install_e2e_site()` which we no longer need:

https://github.com/woocommerce/woocommerce/blob/master/tests/bin/install.sh#L222

I added it to `run-e2e-CI.sh` instead of before running tests in the new Travis implementation.

kloon reviewed (commented):
Just a couple of things I picked up while scanning through the changes visually. More suggestions than anything else. I will go through the testing instructions next and leave feedback on that.

kloon commented on tests/e2e-tests/config/default.json:
```diff
@@ -1,3 +1,13 @@
 {
-  "url": "https://example.com/"
+  "url": "http://example.com",
```
Perhaps we can change this file to have the correct defaults already in place? Eliminates a step to copy and change a file, I like the idea of this file being overridable in case you have a different setup and want to change things a bit though so lets keep that part, just populate this file with the correct default values as defined in the docker files.

juliaamosova replied:
I thought of doing the same and first, have tried it without local `test:e2e.json` file. But when running the tests, I am getting the following warnings in the terminal:

```
console.error node_modules/config/lib/config.js:1425
    WARNING: NODE_ENV value of 'test:e2e' did not match any deployment config file names.

  console.error node_modules/config/lib/config.js:1426
    WARNING: See https://github.com/lorenwest/node-config/wiki/Strict-Mode
```

If I open the [documentation link provided in the warning](https://github.com/lorenwest/node-config/wiki/Strict-Mode), I am getting the following explanation and instructions on how to fix the warning:

> If NODE_ENV is set in the environment, we check if there is any file matching that name in the config/ directory containing valid syntax. You can create a file with the name matching the desired NODE_ENV value: {NODE_ENV}.json and include the contents '{}', which is an empty valid JSON object. If you are using YAML based config files, the file {NODE_ENV}.yml must not be empty, place a comment or any valid YAML syntax in the file.

And so I created `test:e2e.json`. We can technically leave it empty with the contents '{}' - which I tested and it works (given the fact that `default.json` has all the details we need). 

Maybe there is a better way to address the warning and I am missing something simple?

In any case, I also like the idea of keeping `default.json` with the details of the site and user built in Docker. I updated the `default.json` to reflect that.

juliaamosova replied:
I also committed empty `test:e2e.json` so that it could be used in case the default values need to be overridden.

kloon reviewed (changes requested):
On running docker-compose up everything seemed to have run fine, however, I believe there might be a permission error of sort, I got the following warning relating to copying over files and creating directories.

```
wordpress-cli_1             | Success: WordPress installed successfully.
wordpress-cli_1             | Warning: Unable to create directory wp-content/uploads/2019/11. Is its parent directory writable by the server?
wordpress-cli_1             | [26-Nov-2019 13:04:15 UTC] PHP Warning:  copy(/var/www/html/wp-content/uploads/woocommerce-placeholder.png): failed to open stream: No such file or directory in /var/www/html/wp-content/plugins/woocommerce/includes/class-wc-install.php on line 1220
wordpress-cli_1             | Warning: copy(/var/www/html/wp-content/uploads/woocommerce-placeholder.png): failed to open stream: No such file or directory in /var/www/html/wp-content/plugins/woocommerce/includes/class-wc-install.php on line 1220
wordpress-cli_1             | Plugin 'woocommerce' activated.
```
When I navigate to the site url it loads fine though and I can also log in without issues, I suspect we might run into issues when we need to create products and upload images, the current tests does not test images thought so not sure if it is worth fixing then.

Regarding the tests they all passed except one, the Setup Wizard test failed with the following message:
```
Store owner can go through store Setup Wizard › Can fill out Store setup details

    TimeoutError: waiting for selector "#wc-backbone-modal-dialog" failed: timeout 30000ms exceeded

      61 |
      62 | 		// Wait for usage tracking pop-up window to appear
    > 63 | 		await page.waitForSelector( '#wc-backbone-modal-dialog' );
         | 		           ^
      64 |
      65 | 		await page.waitForSelector('#wc_tracker_checkbox_dialog');
      66 |

      at new WaitTask (node_modules/puppeteer/lib/DOMWorld.js:549:28)
      at DOMWorld._waitForSelectorOrXPath (node_modules/puppeteer/lib/DOMWorld.js:478:22)
      at DOMWorld.waitForSelector (node_modules/puppeteer/lib/DOMWorld.js:432:17)
      at Frame.waitForSelector (node_modules/puppeteer/lib/FrameManager.js:627:47)
      at Frame.<anonymous> (node_modules/puppeteer/lib/helper.js:112:23)
      at Page.waitForSelector (node_modules/puppeteer/lib/Page.js:1125:29)
      at Object.waitForSelector (tests/e2e-tests/specs/activate-and-setup/setup-wizard.test.js:63:14)
```

Having kept an eye on the tests as I ran them in dev mode it seems like it was waiting quite a long time on the  payment gateway selection screen on the setup wizard before it actually started selecting anything, I suspect it was waiting for something to pop up which did not.

Other than the above it looks good, great work on this!

juliaamosova committed:
Update docker-compose file & setup running e2e tests on Travis CI

juliaamosova committed:
Update setup-wizard test & WP Docker image version

juliaamosova committed:
Update `default.json` with default Docker test site details

juliaamosova commented:
@kloon thanks for the detailed review. 

>On running docker-compose up everything seemed to have run fine, however, I believe there might be a permission error of sort, I got the following warning relating to copying over files and creating directories.

I see that I am also getting these warnings. And it does look like a permission issue. I am not immediately sure which user should be running creating directories commands with the container yet. I'd need to research a bit more on this. We will likely have an issue with uploading images indeed later on when we'd add it to the tests. 

Regarding the error you are having when running setup wizard test - interestingly, I am not getting these errors. I run these tests multiple times. Nonetheless: 

>TimeoutError: waiting for selector "#wc-backbone-modal-dialog" failed: timeout 30000ms exceeded

`#wc-backbone-modal-dialog` is the selector for the usage tracking pop-up window:

![Screen Shot 2019-11-27 at 12 10 36 AM](https://user-images.githubusercontent.com/19143190/69682753-66004580-10aa-11ea-8ab3-0907578e7345.png)

If it doesn't appear, it is likely that the `Lets go!` button on the Store setup page that triggers the window was not clicked. I updated the way we click on that button here: 

https://github.com/woocommerce/woocommerce/pull/25105/files#diff-1346c6cce433e508ba9f45a142bdd379R60

Can you please test again to see if it works for you now? 

>Having kept an eye on the tests as I ran them in dev mode it seems like it was waiting quite a long time on the payment gateway selection screen on the setup wizard before it actually started selecting anything, I suspect it was waiting for something to pop up which did not.

When we click on the `Continue` button in the usage tracking pop-up window, we need to wait for the Payments page to load. Maybe it was taking a while which is why the test didn't proceed right away. In any case, I also updated the way we click on that button and wait for the payment page to load as per [Puppeteer's documentation](https://github.com/puppeteer/puppeteer/blob/master/docs/api.md#pagewaitfornavigationoptions):

https://github.com/woocommerce/woocommerce/pull/25105/files#diff-1346c6cce433e508ba9f45a142bdd379R71

Can you please test as well to see if it works better for you now? It does work for me in both headless and dev modes. 

I was also wondering if we can test running these tests on Travis to see what we'd get there. In one of my last commits today, I added Travis implementation. I can't test them, however, since this branch was not made of the master. Can we merge it to our migration branch which is off master to see if the Travis implementation works and how tests run there?

juliaamosova committed:
Add `npm install` to the Travis implementation for running e2e tests

juliaamosova commented:
I also combined the 2 commands: stop the container(s) and delete all volumes - to one in the `TO STOP DOCKER & DO A CLEAN RESTART:`:

`docker-compose down -v`

I updated instructions on how to test PR with it.

juliaamosova committed:
Add empty `test:e2e.json` for overriding default variables if needed

peterfabian reviewed (approved):
Really like the direction this is going to. Nice work! 👍 

I've tested it on my setup and all tests have passed, so I'm approving and hoping we'd be able to find out what the problem is with @kloon 's setup.

peterfabian commented on tests/e2e-tests/specs/front-end/front-end-single-product.test.js:
```diff
@@ -19,7 +14,7 @@ let variablePostIdValue;
 describe( 'Single Product Page', () => {
 	beforeAll( async () => {
-		await activatePlugin( 'woocommerce' );
+		await StoreOwnerFlow.login();
```
I like the change from activatePlugin to StoreOwnerFlow, makes more conceptual sense to me.

juliaamosova replied:
👍

peterfabian commented:
Btw this seemed to work for me to get the build working inside docker:

```
NODE_RELEASE=12.x
apt-get update
apt-get install curl
rm -rf ~/.nvm
curl -sL "https://deb.nodesource.com/setup_${NODE_RELEASE}" | bash -
apt-get install -y nodejs git

php -r "copy('https://getcomposer.org/installer', 'composer-setup.php');"
php composer-setup.php --install-dir=/usr/local/bin --filename=composer

composer install --no-dev
npm install

```

juliaamosova commented:
>Btw this seemed to work for me to get the build working inside docker:

Thanks Peter! Let me try it on the migration branch: https://github.com/woocommerce/woocommerce/pull/24218

juliaamosova commented:
I want to document @kloon's findings regarding the commands used to stop Docker and do a clean restart:

The command that copies the files over only runs when it does a clean image creation, even if you do `docker-compose down -v` it still has the image built with the old files already on there. In order to get the fresh files you will have to run `docker-compose up --build` so it does the part again where it copies the files and builds a new container with those files loaded.

With that said, I am updating the `TO STOP DOCKER & DO A CLEAN RESTART:` section of the PR description with the correct command to use.

shivapoudel commented:
@juliaamosova After this PR is merged, I am receiving Invalid Git Paths error on Windows 10 environment. Since Windows won't allow colon in a file name and [this commit](https://github.com/woocommerce/woocommerce/commit/460b2f19e32511472dc3a34381e90ab76096a473) adds such colon based file.
